lib/websocket-server.js

```diff
@@ -151,29 +155,41 @@ class WebSocketServer extends EventEmitter {
     if (this._state === CLOSING) return;
     this._state = CLOSING;
 
-    //
-    // Terminate all associated clients.
-    //
-    if (this.clients) {
-      for (const client of this.clients) client.terminate();
-    }
+    if (this.options.noServer || this.options.server) {
+      if (this._server) {
+        this._removeListeners();
+        this._removeListeners = this._server = null;
+      }
+
+      if (this.clients) {
+        if (!this.clients.size) {
+          process.nextTick(emitClose, this);
+        } else {
+          for (const client of this.clients) client.close(1001);
```

Comment:
From a networking POV, this all seems like a good improvement to me. It definitely seems clearer for clients if servers send a 1001 close frame and tidily shutdown connections where possible :+1:

That said, it does mean the connections could now unexpectedly stay open whilst waiting for a close response, until the close timeout, if you have any poorly behaving clients. That might be problematic if users are trying to _immediately_ shut everything down, which seems like a reasonable use case, and there's now no easy alternative way to do that when necessary.

Maybe the server should have a `server.terminate()` method too, which does terminate everything immediately?

That would make the server API match the individual websockets API, which already has `websocket.terminate()`, and it would cover that immediate-shutdown use case.

As a bonus, this could also easily support cases like "try to shut down cleanly, but then always close within X milliseconds" with just:

```javascript
server.close();
setTimeout(() => server.terminate(), X);
```

Reply:
Users can terminate clients by themself if they want to close immediately:

```js
wss.close();
for (const ws of wss.clients) client.terminate();
```

Reply:
Ah ok, fair enough. I'm not sure if that trick will be obvious to people, so a convenience `server.terminate()` method to do the same could still be useful, but you're right that that means it's not strictly necessary :+1:

It's probably worth mentioning this in the breaking-changes release notes though imo, as the suggested way to migrate for people who do need the previous behaviour.

Reply:
I have two doubts:

1. If it is better to just prevent new connections from being established and not close/terminate existing clients. Closing clients is user responsibility so they can use `websocket.close()` or `websocket.terminate()` at will. This is also consistent with the behavior of [`server.close()`](https://nodejs.org/api/net.html#net_server_close_callback)
1. Regardless of 1., if/when to emit the `'close'` event when client tracking is disabled. This is currently done in the next tick.

Reply:
I went ahead and changed `WebSocketServer.prototype.close()` behavior in https://github.com/websockets/ws/pull/1904/commits/6a238da218d44ef4e68177dae1d54280bd8eeee8. It no longer closes existing connections as per point 1. of my previous comment.

I've also updated the documentation for point 2. I think emitting `'close'` when connections are still active is a bit misleading but this behavior already exists and was not changed here.

Reply:
Missed this sorry, I've been busy. That sounds very reasonable to me and the consistency with node is great.

I do think documenting this clearly for those who do want to terminate the server will be useful though, and it's not obvious right now. It might be worth a "how to fully shutdown a server" section in the readme, and/or the release notes.